### PR TITLE
Support getting eligible replicas by state in OperationTracker

### PR DESCRIPTION
**Pull request by jsjtzyy**
Previously, OperationTracker directly calls partition.getReplicaIds() to
populate replicaPool before sending out requests. With this change
OperationTracker is able to get eligible replicas by state (i.e. STANDBY,
LEADER) and then populate replicaPool with eligible replicas only.
Introducing eligible replicas will help router avoid sending requests to
some replicas in BOOTSTRAP or INACTIVE state due to ongoing replica movement.

### CODE REVIEW

**Comment by jsjtzyy** [comment]
@zzmao @justinlin-linkedin I fixed the test failure and the PR is ready for review.

File: ambry-router/src/test/java/com.github.ambry.router/OperationTrackerTest.java [hunk 1]
```diff
@@ -152,6 +159,133 @@ public void localFailTest() {
...
+    populateReplicaList(2, ReplicaState.STANDBY);
+    localDcName = datanodes.get(0).getDatacenterName();
+    mockClusterMap = new MockClusterMap(false, datanodes, 1, Collections.singletonList(mockPartition), localDcName);
+    OperationTracker ot = getOperationTracker(true, 1, 2, true, Integer.MAX_VALUE, RouterOperation.PutOperation);
+    assertFalse("Operation should not have been done.", ot.isDone());
+    sendRequests(ot, 2, false);
+    // make one requests succeed, the other fail
+    ot.onResponse(inflightReplicas.poll(), TrackedRequestFinalState.SUCCESS);
+    ot.onResponse(inflightReplicas.poll(), TrackedRequestFinalState.FAILURE);
+    assertFalse("Operation should fail", ot.hasSucceeded());
```
Comment (justinlin-linkedin) [inline ambry-router/src/test/java/com.github.ambry.router/OperationTrackerTest.java]: the successTarget is 1 here, why would it fail?

File: ambry-router/src/main/java/com.github.ambry.router/SimpleOperationTracker.java [hunk 2]
```diff
@@ -84,13 +89,35 @@
...
+   * |     GET          | STANDBY, LEADER, BOOTSTRAP, INACTIVE |
+   * |    DELETE        | STANDBY, LEADER, BOOTSTRAP           |
+   * |   TTLUpdate      | STANDBY, LEADER, BOOTSTRAP           |
+   *  ---------------------------------------------------------
+   * Following are dynamic configs when replica state is taken into consideration: (N is number of eligible replicas)
+   *  -----------------------------------------------------------------------
+   * |  Operation Type  |        Parallelism              |  Success Target  |
+   *  -----------------------------------------------------------------------
+   * |     POST         | 1~2 decided by adaptive tracker |         1        |
+   * |     GET          |           N                     |       N - 1      |
```
[the remark this answers is not in the document]
Comment (jsjtzyy) [reply ambry-router/src/main/java/com.github.ambry.router/SimpleOperationTracker.java]: right, will fix

File: ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockPartitionId.java [hunk 3]
```diff
@@ -71,7 +77,13 @@ public MockPartitionId(long partition, String partitionClass, List<MockDataNodeI
 
   @Override
   public List<ReplicaId> getReplicaIdsByState(ReplicaState state, String dcName) {
-    throw new UnsupportedOperationException("Temporarily Unsupported");
+    List<ReplicaId> replicas = new ArrayList<>();
+    replicaIds.forEach(r -> {
+      if (replicaAndState.get(r) == state && (dcName == null || r.getDataNodeId().getDatacenterName().equals(dcName))) {
+        replicas.add(r);
+      }
+    });
```
[the remark this answers is not in the document]
Comment (jsjtzyy) [reply ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockPartitionId.java]: sure, will fix

File: ambry-router/src/test/java/com.github.ambry.router/OperationTrackerTest.java (hunk 1 above)
Comment (jsjtzyy) [reply ambry-router/src/test/java/com.github.ambry.router/OperationTrackerTest.java]: successTarget is 2 for PUT operation. This test is for PUT scenario.

File: ambry-router/src/main/java/com.github.ambry.router/SimpleOperationTracker.java [hunk 4]
```diff
@@ -76,9 +80,32 @@
...
   /**
-   * Constructor for an {@code SimpleOperationTracker}.
+   * Constructor for an {@code SimpleOperationTracker}. In constructor, there is a config allowing operation tracker to
+   * use eligible replicas to populate replica pool. ("eligible" replicas are those in required states for specific
+   * operation)
+   * Following are different types of operation and their eligible replica states:
+   *  ---------------------------------------------------------
+   * |  Operation Type  |        Eligible Replica State        |
+   *  ---------------------------------------------------------
+   * |     POST         | STANDBY, LEADER                      |
```
Comment (justinlin-linkedin) [inline ambry-router/src/main/java/com.github.ambry.router/SimpleOperationTracker.java]: This should be PUT also

Comment (jsjtzyy) [reply ambry-router/src/main/java/com.github.ambry.router/SimpleOperationTracker.java]: oops, will fix

File: ambry-api/src/main/java/com.github.ambry/config/RouterConfig.java [hunk 5]
```diff
@@ -364,10 +364,20 @@
...
 
+  /**
+   * If {@code true}, the blob data shares memory with networking buffer in GetBlobOperation
+   */
   @Config("router.get.blob.operation.share.memory")
   @Default("false")
   public final boolean routerGetBlobOperationShareMemory;
 
+  /**
+   * if {@code true}, operation tracker will get replicas in required states based on the type of operation.
```
Comment (zzmao) [inline ambry-api/src/main/java/com.github.ambry/config/RouterConfig.java]: add comment ``` This is used for dynamic replica add/remove```  would be more friendly.

**Pull request merged**